### PR TITLE
feat: Implement stylelint plugin

### PR DESCRIPTION
**Pull request by ling1726**
Implements a stylelint plugin for Fluent UI specific rules. Starts off
with 2 rules:

* combinator-depth
* no-focus-visible

### CODE REVIEW

File: packages/stylelint-plugin/src/create-rule.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { Rule, RuleMessages, RuleBase } from 'stylelint';
```
Comment (layershifter) [inline packages/stylelint-plugin/src/create-rule.ts:1]:
```suggestion
import type { Rule, RuleMessages, RuleBase } from 'stylelint';
```

File: packages/stylelint-plugin/src/postcss-utils.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import type { Document, AtRule, Root, Container, Rule } from 'postcss';
+export function isAtRule(container: { type: string }): container is AtRule {
```
Comment (layershifter) [inline packages/stylelint-plugin/src/postcss-utils.ts:2]:
```suggestion
import type { Document, AtRule, Root, Container, Rule } from 'postcss';

export function isAtRule(container: { type: string }): container is AtRule {
```

File: packages/stylelint-plugin/src/rules/combinator-depth/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+id or a CSS class directly on the affected DOM element.
+
+## Incorrect
+
+```tsx
+import { makeStyles, Button, buttonClassNames } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  root: {
+    [`.${buttonClassNames.root}`]: {
```
Comment (layershifter) [inline packages/stylelint-plugin/src/rules/combinator-depth/README.md:14]:
```suggestion
    [`& .${buttonClassNames.root}`]: {
```

File: packages/stylelint-plugin/src/rules/combinator-depth/combinator-depth.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      postcssResult,
+      ruleName,
+      {
+        actual: allowedDepth,
+        possible: [(val) => typeof val === 'number'],
+      }
+    );
+
+    if (!validOptions) {
+      return;
```
Comment (layershifter) [inline packages/stylelint-plugin/src/rules/combinator-depth/combinator-depth.ts:30]: Should not it throw?

File: packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.test.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
+import rule from './no-focus-visible';
+import * as stylelint from 'stylelint';
+
+describe('no-focus-visible', () => {
+  it('should fail with :focus-visible', async () => {
+    const { errored, results } = await stylelint.lint({
+      code: `.foo:focus-visible { color: red; }`,
```
Comment (layershifter) [inline packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.test.ts:7]: Let's add a fixture with `.foo :focus-visible` to ensure that space is handled, too.

File: packages/stylelint-plugin/src/rules/combinator-depth/combinator-depth.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        continue;
+      }
+      tokenizedSelector.reverse();
+      let combinatorCount = 0;
+      for (const token of tokenizedSelector) {
+        if (token.type === 'combinator') {
+          combinatorCount++;
+        }
+      }
```
Comment (layershifter) [inline packages/stylelint-plugin/src/rules/combinator-depth/combinator-depth.ts:55]:
```suggestion
      const combinatorCount = tokenizedSelector.filter(token.type === 'combinator').length;
```

?

File: packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        )
+      ) {
+        stylelint.utils.report({
+          ruleName,
+          message: messages.focusVisible(selector),
+          messageArgs: [selector],
+          result: postcssResult,
+          node: selectorToRule[selector],
+        });
+      } else if (
```
Comment (layershifter) [inline packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.ts:59]:
```suggestion
        });
        continue;
      }
      
      if (
```

File: packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    if (!validOptions) {
+      return;
+    }
+
+    const { selectorToRule, selectors } = extractAllSelectors(postcssRoot);
+
+    for (const selector of selectors) {
+      let tokenizedSelector = [];
+      try {
+        tokenizedSelector = parsel.tokenize(selector);
```
Comment (layershifter) [inline packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.ts:34]: Do you really need to tokenize there? Can't we simply use `.indexOf(":focus-visible")`?

**Review by layershifter — APPROVED** [review]

File: packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.ts (hunk 7 above)
Comment (ling1726) [reply packages/stylelint-plugin/src/rules/no-focus-visible/no-focus-visible.ts:59]: personally I prefer `else if` over `continue`

**Pull request merged**